src/main/java/org/reaktivity/nukleus/kafka/internal/stream/ClientStreamFactory.java

```diff
@@ -230,19 +242,28 @@ private MessageConsumer newConnectReplyStream(
     private RouteFW resolveRoute(
         long authorization,
         long sourceRef,
-        String topicName)
+        final String topicName,
+        final ListFW<KafkaHeaderFW> headers)
     {
         final MessagePredicate filter = (t, b, o, l) ->
         {
             final RouteFW route = routeRO.wrap(b, o, l);
             final OctetsFW extension = route.extension();
             Predicate<String> topicMatch = s -> true;
+            Predicate<ListFW<KafkaHeaderFW>> headersMatch = h -> true;
             if (extension.sizeof() > 0)
             {
                 final KafkaRouteExFW routeEx = extension.get(routeExRO::wrap);
                 topicMatch = routeEx.topicName().asString()::equals;
+                if (routeEx.headers().sizeof() > 0)
+                {
+                    headersMatch = candidate ->
+                            !routeEx.headers().anyMatch(r -> !candidate.anyMatch(
+                                    h -> h.key().asString().equals(r.key().asString())
+                                    && BufferUtil.matches(r.value(),  h.value())));
+                }
             }
-            return route.sourceRef() == sourceRef && topicMatch.test(topicName);
+            return route.sourceRef() == sourceRef && topicMatch.test(topicName) && headersMatch.test(headers);
```

Comment:
Verify behavior for `BEGIN` with headers `header1="one"`, `header2="two"`, `header3="three"` and `ROUTE` with headers `header1="one"`, `header2="two"` only. I think we wanted that to match, then filter matching messages for delivery.

Reply:
done (at least similar: route with header1=value1, begin with header1=value1, header2=value2)